flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,621 @@
+The copied capability will use the link on the same path as the original capability.
+
+## Problem statement
+
+The current capability API has a few pain points:
+
+- The concepts capabilities/linking/revoking/redirecting are hard to understand for new developers that are coming to Flow,
+as that is a lot of new concepts to grasp before one can get started.
+
+- Issuing and managing private capabilities that can be revoked at a granular level,
+  by creating a custom private linked path for each capability, is difficult.
+
+- Unless the path name includes some hints, there is no way to know when a path was created,
+  thus making it more difficult to remember why a certain capability was issued.
+
+- If an old path (that has been unlinked) is accidentally reused and re-linked,
+  this will revive a capability that is supposed to be revoked.
+
+## Suggested change
+
+The suggested change addresses these pain points by:
+
+- Removing (or abstracting away) the concept of links.
+- Making it easier to create new capabilities (with individual links).
+- Offering a way to iterate through capabilities on a storage path.
+- Removing the need to have a `/private` domain.
+- Introducing Capability Controllers (CapCons) that handle management of capabilities.
+
+### Capability Controllers (CapCons)
+
+Each Capability has an associated CapCon that is responsible for managing the Capability.
+The CapCon is created when the Capability is issued.
+When the Capability is copied (since it is a value type),
+it shares the CapCon with the original capability
+(i.e., calling `delete` on the CapCon revokes all copies of the associated capability).
+
+The Capability, and all copies of that Capability, have the same ID.
+Capability IDs are unique per account.
+
+CapCons are not storable (similar to e.g. AuthAccounts).
+
+There are two kinds of capabilities:
+- Storage capability: Targets a storage path in an account
+- Account capabilities: Targets an account.
+  This kind was introduced in [FLIP 53](https://github.com/onflow/flips/pull/53).
+
+Therefore, there are two kinds of Capability Controllers:
+`StorageCapabilityController` and `AccountCapabilityController`.
+
+```cadence
+pub struct StorageCapabilityController {
+
+    /// The type of the controlled capability, i.e. the T in `Capability<T>`.
+    pub let borrowType: Type
+
+    /// The identifier of the controlled capability.
+    /// All copies of a capability have the same ID.
+    pub let capabilityID: UInt64
+
+    /// Delete this capability controller,
+    /// and disable the controlled capability and its copies.
+    ///
+    /// The controller will be deleted from storage,
+    /// but the controlled capability and its copies remain.
+    ///
+    /// Once this function returns, the controller is no longer usable,
+    /// all further operations on the controller will panic.
+    ///
+    /// Borrowing from the controlled capability or its copies will return nil.
+    ///
+    pub fun delete()
```

Comment:
Above, you say that capcons aren't storable, but you say this deletes it from storage. I'm a little confused what this is referring to. The same comment applies to the one for the account capability controller as well

Reply:
Good point, I'll clarify the wording. 

Capability controllers are stored in the account's storage, but are not a first-class value that can be e.g. stored inside another value or saved into account storage using `save` – the execution environment manages the creation, storage, and deletion itself.

Initially, the FLIP proposed that capability controllers could be accessed as a value, i.e. a program would work with e.g a `StorageCapabilityController`, similar to how transactions work with `AuthAccount` values.
However, that would have meant that a user could attempt to e.g. store it inside of another value or store it into the account storage with the `save` function. For this purpose, Cadence as a built-in notion of "Storable", and e.g. `AuthAccount` is not a storable type.

However, we can simply enforce the same restriction by not giving out full values of the controller, but just a reference – references are not storable.

Reply:
Clarified this in 0c34ed40464b162ba9a6f85291ff3ab4ef953935